component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/smsotp/SMSOTPAuthenticator.java

```diff
@@ -777,42 +780,64 @@ private void processValidUserToken(AuthenticationContext context, AuthenticatedU
     private void checkWithBackUpCodes(AuthenticationContext context, String userToken,
                                       AuthenticatedUser authenticatedUser) throws AuthenticationFailedException {
 
-        String savedOTPString = null;
+        String[] savedOTPs = null;
         String username = context.getProperty(SMSOTPConstants.USER_NAME).toString();
         String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
         UserRealm userRealm = getUserRealm(username);
         try {
             if (userRealm != null) {
-                savedOTPString = userRealm.getUserStoreManager()
-                        .getUserClaimValue(tenantAwareUsername, SMSOTPConstants.SAVED_OTP_LIST, null);
-            }
-            if (StringUtils.isEmpty(savedOTPString)) {
-                if (log.isDebugEnabled()) {
-                    log.debug("The claim " + SMSOTPConstants.SAVED_OTP_LIST + " does not contain any values");
-                }
-                throw new AuthenticationFailedException("The claim " + SMSOTPConstants.SAVED_OTP_LIST +
-                        " does not contain any values");
-            } else if (savedOTPString.contains(userToken)) {
-                if (log.isDebugEnabled()) {
-                    log.debug("Found saved backup SMS OTP for user :" + authenticatedUser);
+                UserStoreManager userStoreManager = userRealm.getUserStoreManager();
+                if (userStoreManager != null) {
+                    String savedOTPString = userStoreManager
+                            .getUserClaimValue(tenantAwareUsername, SMSOTPConstants.SAVED_OTP_LIST, null);
+                    if (StringUtils.isNotEmpty(savedOTPString)) {
+                        savedOTPs = savedOTPString.split(FrameworkUtils.getMultiAttributeSeparator());
+                    }
                 }
-                context.setSubject(authenticatedUser);
-                savedOTPString = savedOTPString.replaceAll(userToken, "").replaceAll(",,", ",");
-                userRealm.getUserStoreManager().setUserClaimValue(tenantAwareUsername,
-                        SMSOTPConstants.SAVED_OTP_LIST, savedOTPString, null);
-            } else {
-                if (log.isDebugEnabled()) {
-                    log.debug("User entered OTP :" + userToken + " does not match with any of the saved backup codes");
+            }
+            if (!ArrayUtils.isEmpty(savedOTPs)) {
+                if (savedOTPs.length == 0) {
```

Comment:
ArrayUtils.isEmpty() will check the length == 0 as well. So this check is reduent